src/avram/nested_save_operation.cr

```diff
@@ -1,35 +1,85 @@
 module Avram::NestedSaveOperation
-  macro has_one(type_declaration)
-    {% name = type_declaration.var %}
-    {% type = type_declaration.type %}
-    @_{{ name }} : {{ type }} | Nil
-
-    def {{ name }}
-      @_{{ name }} ||= {{ type }}.new(params)
+  macro included
+    macro inherited
+      define_has_one
     end
+  end
+
+  macro define_has_one
+    macro has_one(type_declaration)
+      \{% name = type_declaration.var %}
+      \{% type = type_declaration.type.resolve %}
+
+      \{% model_type = type.ancestors.find do |t|
+          t.stringify.starts_with?("Avram::SaveOperation(")
+        end.type_vars.first %}
+
+      \{% assoc = T.constant(:ASSOCIATIONS).find do |assoc|
+          assoc[:relationship_type] == :has_one &&
+            assoc[:type].resolve.name == model_type.name
+        end %}
+
+      \{% if ! assoc %}
+        \{% raise "#{T} must have a has_one association with #{model_type}" %}
+      \{% end %}
+
+      \{% for need in type.constant(:OPERATION_NEEDS) %}
+        needs \{{ name }}_\{{ need }}
+      \{% end %}
 
-    after_save save_nested_{{ name }}
+      after_save save_\{{ name }}
 
-    def save_nested_{{ name }}(record)
-      {{ name }}.{{ @type.constant(:FOREIGN_KEY).id }}.value = record.id
+      def save_\{{ name }}(saved_record)
+        unless \{{ name }}.save
+          add_error(:\{{ name }}, "failed")
+          mark_nested_save_operations_as_failed
+          database.rollback
+        end
+      end
+
+      def \{{ name }}
+        @\{{ name }} ||= if new_record?
+          \{% begin %}\{{ type }}.new(
+            params,
+            \{% for need in type.constant(:OPERATION_NEEDS) %}
+              \{{ need.var }}: \{{ name }}_\{{ need.var }},
+            \{% end %}
+          )\{% end %}
+        else
+          \{% begin %}\{{ type }}.new(
+            record.not_nil!.\{{ assoc[:assoc_name].id }}!,
+            params,
+            \{% for need in type.constant(:OPERATION_NEEDS) %}
+              \{{ need.var }}: \{{ name }}_\{{ need.var }},
+            \{% end %}
+          )\{% end %}
+        end
+
+        nested = @\{{ name }}.not_nil!
+
+        record.try do |record|
+          nested.\{{ @type.constant(:FOREIGN_KEY).id }}.value = record.id
+        end
```

Comment:
This should probably happen in the `after_save` call. You may call the association during some before_saves and it keeps trying to make this assignment on an accessor method.